benchmarks/python/optimized-numpy.py

```diff
@@ -0,0 +1,123 @@
+import sys
+import math
+import timeit
+
+import numpy as np
+import pandas as pd
+
+def load_input_data(path):
+
+    df = pd.read_csv(
+        path, names = ["mass", "x", "y", "z", "vx", "vy", "vz"], delimiter = r"\s+"
+    )
+
+    masses = df["mass"].values.copy()
+    positions = df.loc[:, ["x", "y", "z"]].values.copy()
+    velocities = df.loc[:, ["vx", "vy", "vz"]].values.copy()
+
+    return masses, positions, velocities
+
+def compute_accelerations(accelerations, masses, positions):
+    nb_particles = masses.size
+
+    for index_p0 in range(nb_particles - 1):
+        position0 = positions[index_p0]
+        mass0 = masses[index_p0]
+
+        vectors = position0 - positions[index_p0 + 1: nb_particles]
+
+        distances = np.square(vectors).sum(axis = 1)
+        coefs = distances ** 1.5
+
+        _x = np.add(np.sum(
+            np.divide(
+                np.multiply(
+                    masses[index_p0 + 1: nb_particles], -1 * vectors.T
+                    ),
+                coefs).T, axis = 0),
+            accelerations[index_p0]
+            )
```

Comment:
Hi,
I agree with you that we can make the code simpler. I tried to do so, especially at this part. I observed a few things which I would like to share and get your inputs. 

I tried the following ways to calculate accelerations:

- The part I had updated in PR.
    - The time it took: `187.2880`.
- When I changed that part too:
  ```
  ind0_mul = masses[index_p0 + 1: nb_particles] * (-1 * vectors.T)
  ind0_div = ind0_mul / coefs
  ind0_div = ind0_div.T
  ind0_sum = np.sum(ind0_div, axis = 0)
  accelerations[index_p0] += accelerations[index_p0]
  
  ind1_mul = mass0 * vectors.T
  ind1_div = ind1_mul / coefs
  ind1_div = ind1_div.T
  accelerations[index_p0 + 1: nb_particles] += ind1_div
  ```
   - The time it took: `196.3231`.
- Another way I tried is the simplest to understand. But does not perform well is.
  ```
  def compute_accelerations(accelerations, masses, positions):
      nb_particles = masses.size
  
      for index_p0 in range(nb_particles - 1):
          position0 = positions[index_p0]
          mass0 = masses[index_p0]
  
          vector = np.empty(3)
  
          for index_p1 in range(index_p0 + 1, nb_particles):
              mass1 = masses[index_p1]
              position1 = positions[index_p1]
  
              for i in range(3):
                  vector[i] = position0[i] - position1[i]
              
              distance = sum(vector ** 2) * math.sqrt(sum(vector ** 2))
              coef_m1 = mass0 / distance
              coef_m2 = mass1 / distance
  
              for i in range(3):
                  accelerations[index_p0][i] += coef_m1 * -1 * vector[i]
                  accelerations[index_p1][i] += coef_m2 * vector[i]
  
      return accelerations
  ```
  - The time it took: `475.0748`.

#### Prerequisite
- These results are for `50` iterations, for datasets of input size `16` at an interval of `10` seconds, after each `0.001` step.
- **Command to run the files**:
   ```
    taskset -c 6,7,14,15 python python/optimized-numpy.py data/input16.txt 10.0
  ```
  Here, 
  - `taskset -c 6,7,14,15`: for running files in isolated cpu's.
  - `python/optimized-numpy.py`: python file, whose location is in the python folder.
  - `data/input16.txt`: denotes dataset.
  - `10.0`: total time to iterate.
- **Machine Configuration**:
  - OS: `ManjaroLinux 21.1.1`
  - Model Name: `Intel(R) Core(TM) i7-10870H CPU @ 2.20GHz`
  - RAM: `16GB`
- **Dependencies**:
  - Python: `3.9.6`
  - NumPy: `1.20.3`
  - Pythran: `0.9.12.post1`
  - Transonic: `0.4.10`
  - Numba: `0.54.0`
  - CMake: `3.21.1`
  - GCC: `11.1.0`

Reply:
The last version is nice, cooking it down to proper numpy seems not so bad (i.e. remove for loops, except possibly the outer one), then consider using in-place logic.

My main criticism is that it feels you do different operations compared to C++, and that changes the theoretical work done by the programs.

For example:
```
ind0_mul = masses[index_p0 + 1: nb_particles] * (-1 * vectors.T)
ind0_div = ind0_mul / coefs
```
vs.:
```
coef_m1 = mass0 / distance
``
reorders the multiplication with vectors.  The reorder itself is fine (although it would be nicer to avoid for comparison?).  But `ind0_mul` already includes the `for i in range(3):` loop (and maybe more?) so it has to do 3 times as many divisions.

I see the divisions, because they are relatively slow.  Of course this may have nothing to do with the timings you observe.  A profiler (line profiler) might help understand them easier.

Another example, is that I bet the C++ code does not do `-1 * vectors`, but rather uses `acceleration -= ...` at the end, which avoids one operation (albeit probably one that just doesn't matter).

Reply:
Hi, 
Thanks for the brief description. As suggested earlier, I used Python operators and removed the for-loop. Also, I modified the code to maintain uniformity between them. 
Yes! Now it's more readable, and the performance of the model is also improved. 

> I see the divisions, because they are relatively slow. Of course this may have nothing to do with the timings you observe. A profiler (line profiler) might help understand them easier.

Yes, divisions are really slow. I had modified it and updated the PR. 

> Another example, is that I bet the C++ code does not do -1 * vectors, but rather uses acceleration -= ... at the end, which avoids one operation

Indeed, You won the bet!!

But  I have a different kind of opinion on this part and would like to discuss it. I agree that using `acceleration -= ...` avoids one operation that might significantly increase the performance of the model for larger datasets. At the same time,
Let us consider two bodies at a distance, `d`, under mutual gravitational force of attraction. The 1st body is moving from position `i` to position `j`. Hence, the position vector of the 1st body will be represented as `r_j - r_i`. And the 2nd body is moving from position `j` to position `i`. Its position vector is represented as `r_i - r_j`. 
Therefore I concluded that it is conceptually more strong to use `-1 * vectors` than to use `acceleration -= ...`. Though, the results are the same. Does that make sense? What do you think?

Reply:
The following are the results of line_profiler from a few methods I tried:

- The best one (till now):
  ```
  time taken: 7.4037153270001
  Wrote profile results to optimized_numpy5.py.lprof
  Timer unit: 1e-06 s
  
  Total time: 6.60908 s
  File: python/optimized_numpy5.py
  Function: compute_accelerations at line 20
  
  Line #      Hits         Time  Per Hit   % Time  Line Contents
  ==============================================================
      20                                           @profile
      21                                           def compute_accelerations(accelerations, masses, positions):
      22     10002       7759.0      0.8      0.1      nb_particles = masses.size
      23                                           
      24    160032     103357.0      0.6      1.6      for index_p0 in range(nb_particles - 1):
      25    150030     126036.0      0.8      1.9          position0 = positions[index_p0]
      26    150030     101733.0      0.7      1.5          mass0 = masses[index_p0]
      27                                           
      28    150030     398855.0      2.7      6.0          vectors = position0 - positions[index_p0 + 1: nb_particles]
      29                                           
      30    150030    1057577.0      7.0     16.0          distances = (vectors**2).sum(axis=1)
      31    150030     468228.0      3.1      7.1          coefs = distances**1.5
      32                                           
      33    150030    3144079.0     21.0     47.6          accelerations[index_p0].__iadd__(np.sum((masses[index_p0 + 1: nb_particles] * -1 * vectors.T / coefs).T, axis=0))
      34    150030    1197118.0      8.0     18.1          accelerations[index_p0 + 1: nb_particles].__iadd__((mass0 * vectors.T / coefs).T)
      35                                           
      36     10002       4341.0      0.4      0.1      return accelerations
  ```

- Without using in-place logic:
  ```
  35    150030    2897902.0     19.3     46.3          accelerations[index_p0] += np.sum((masses[index_p0 + 1: nb_particles] * -1 * vectors.T * coefs).T, axis=0)
  36    150030    1083315.0      7.2     17.3          accelerations[index_p0 + 1: nb_particles] += (mass0 * vectors.T * coefs).T
  ```

- When I changed some method to calculate distances:
  ```
  30    150030    1938345.0     12.9     27.3          distances = np.sum((vectors**2), axis=1) #.sum(axis=1)
  31    150030     601801.0      4.0      8.5          coefs = 1./distances**1.5
  ```
- Another way of the same:
  ```
  32    150030     806488.0      5.4     12.5          coefs = 1./np.sqrt(distances**3)
  33    150030    2879604.0     19.2     44.6          accelerations[index_p0].__iadd__(np.sum((masses[index_p0 + 1: nb_particles] * -1 * vectors.T * coefs).T, axis=0))
  34    150030    1064629.0      7.1     16.5          accelerations[index_p0 + 1: nb_particles].__iadd__((mass0 * vectors.T * coefs).T)
  ```
- The most important, that I used earlier (this was really very slow, comparatively):
  ```
  31    150030     424432.0      2.8      7.1          coefs = distances**1.5
  32                                                   #coefs = 1./distances**1.5
  33                                                   #coefs = 1./np.sqrt(distances**3)
  34    150030    2849387.0     19.0     47.6          accelerations[index_p0].__iadd__(np.sum((masses[index_p0 + 1: nb_particles] * -1 * vectors.T / coefs).T, axis=0))
  35    150030    1070838.0      7.1     17.9          accelerations[index_p0 + 1: nb_particles].__iadd__((mass0 * vectors.T / coefs).T)
  ```